pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/headroompolicy/policy_adaptive.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2022 The Katalyst Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package headroompolicy
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/kubewharf/katalyst-api/pkg/consts"
+	"github.com/kubewharf/katalyst-core/pkg/agent/qrm-plugins/cpu/dynamicpolicy/state"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/metacache"
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/types"
+	"github.com/kubewharf/katalyst-core/pkg/config"
+	"github.com/kubewharf/katalyst-core/pkg/config/agent/sysadvisor/qosaware/resource/cpu/headroom"
+	pkgconsts "github.com/kubewharf/katalyst-core/pkg/consts"
+	"github.com/kubewharf/katalyst-core/pkg/metaserver"
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+	"github.com/kubewharf/katalyst-core/pkg/util/metric"
+)
+
+type PolicyAdaptive struct {
+	*PolicyBase
+
+	policyAdaptiveConfig *headroom.PolicyAdaptiveConfiguration
+}
+
+func NewPolicyAdaptive(regionName string, conf *config.Configuration, _ interface{}, metaReader metacache.MetaReader,
+	metaServer *metaserver.MetaServer, emitter metrics.MetricEmitter) HeadroomPolicy {
+	p := &PolicyAdaptive{
+		PolicyBase:           NewPolicyBase(regionName, metaReader, metaServer, emitter),
+		policyAdaptiveConfig: conf.CPUHeadroomPolicyConfiguration.PolicyAdaptive,
+	}
+
+	return p
+}
+
+func (p *PolicyAdaptive) Update() error {
+	regionInfo, ok := p.metaReader.GetRegionInfo(p.regionName)
```

Comment:
why we need to get region-info and judge with its type? should we just verify this logic when NewPolicyAdaptive is called? for dedicated_cores, can it still work since it will never update() with nil error at all?